pkg/report/predicate/vuln_test.go

```diff
@@ -0,0 +1,113 @@
+package predicate_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"testing"
+	"time"
+
+	dbTypes "github.com/aquasecurity/trivy-db/pkg/types"
+	"github.com/aquasecurity/trivy-db/pkg/vulnsrc/vulnerability"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/aquasecurity/trivy/pkg/clock"
+	ftypes "github.com/aquasecurity/trivy/pkg/fanal/types"
+	"github.com/aquasecurity/trivy/pkg/report/predicate"
+	"github.com/aquasecurity/trivy/pkg/types"
+)
+
+func TestWriter_Write(t *testing.T) {
+	testCases := []struct {
+		name          string
+		detectedVulns []types.DetectedVulnerability
+		want          predicate.CosignVulnPredicate
+		wantResult    types.Report
+	}{
+		{
+			name: "happy path",
+			detectedVulns: []types.DetectedVulnerability{
+				{
+					VulnerabilityID:  "CVE-2020-0001",
+					PkgName:          "foo",
+					InstalledVersion: "1.2.3",
+					FixedVersion:     "3.4.5",
+					PrimaryURL:       "https://avd.aquasec.com/nvd/cve-2020-0001",
+					Vulnerability: dbTypes.Vulnerability{
+						Title:       "foobar",
+						Description: "baz",
+						Severity:    "HIGH",
+						VendorSeverity: map[dbTypes.SourceID]dbTypes.Severity{
+							vulnerability.NVD: dbTypes.SeverityHigh,
+						},
+					},
+				},
+			},
+			want: predicate.CosignVulnPredicate{
+				Scanner: predicate.Scanner{
+					URI:     "pkg:github/aquasecurity/trivy@dev",
+					Version: "dev",
+					Result: types.Report{
+						SchemaVersion: 2,
+						ArtifactName:  "alpine:3.14",
+						ArtifactType:  ftypes.ArtifactType(""),
+						Metadata:      types.Metadata{},
+						Results: types.Results{
+							{
+								Target: "foojson",
+								Vulnerabilities: []types.DetectedVulnerability{
+									{
+										VulnerabilityID:  "CVE-2020-0001",
+										PkgName:          "foo",
+										InstalledVersion: "1.2.3",
+										FixedVersion:     "3.4.5",
+										PrimaryURL:       "https://avd.aquasec.com/nvd/cve-2020-0001",
+										Vulnerability: dbTypes.Vulnerability{
+											Title:       "foobar",
+											Description: "baz",
+											Severity:    "HIGH",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+				Metadata: predicate.Metadata{
+					ScanStartedOn:  time.Date(2022, time.July, 22, 12, 20, 30, 5, time.UTC),
+					ScanFinishedOn: time.Date(2022, time.July, 22, 12, 20, 30, 5, time.UTC),
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+
+			inputResults := types.Report{
+				SchemaVersion: 2,
+				ArtifactName:  "alpine:3.14",
+				Results: types.Results{
+					{
+						Target:          "foojson",
+						Vulnerabilities: tc.detectedVulns,
+					},
+				},
+			}
+
+			output := bytes.NewBuffer(nil)
+
+			clock.SetFakeTime(t, time.Date(2022, 7, 22, 12, 20, 30, 5, time.UTC))
+			writer := predicate.NewWriter(output, "dev")
+
+			err := writer.Write(inputResults)
+			assert.NoError(t, err)
```

Comment:
nit: IMHO, `require` is better here as all the subsequent assertions will fail anyway when it returns an error.

```suggestion
			require.NoError(t, err)
```

Reply:
It makes sense. I've fixed it.